dll/win32/iphlpapi/icmp.c

```diff
@@ -1,964 +1,339 @@
+ *		IcmpCreateFile (IPHLPAPI.@)
  */
-static LONG icmp_sequence=0;
-
-static int in_cksum(u_short *addr, int len)
+HANDLE WINAPI IcmpCreateFile(VOID)
 {
-    int nleft=len;
-    u_short *w = addr;
-    int sum = 0;
-    u_short answer = 0;
-
-    while (nleft > 1) {
-        sum += *w++;
-        nleft -= 2;
-    }
-
-    if (nleft == 1) {
-        *(u_char *)(&answer) = *(u_char *)w;
-        sum += answer;
+    OBJECT_ATTRIBUTES ObjectAttributes;
+    IO_STATUS_BLOCK IoStatusBlock;
+    UNICODE_STRING DeviceName;
+    HANDLE FileHandle;
+    NTSTATUS Status;
+    DWORD Error;
+
+    /* ICMP requests are handled by the Ip device */
+    RtlInitUnicodeString(&DeviceName, L"\\Device\\Ip");
+    InitializeObjectAttributes(&ObjectAttributes,
+                               &DeviceName,
+                               OBJ_CASE_INSENSITIVE,
+                               NULL,
+                               NULL);
+
+    /* Get a handle to the device */
+    Status = NtCreateFile(&FileHandle,
+                          GENERIC_EXECUTE,
+                          &ObjectAttributes,
+                          &IoStatusBlock,
+                          0,
+                          FILE_ATTRIBUTE_NORMAL,
+                          FILE_SHARE_READ | FILE_SHARE_WRITE,
+                          FILE_OPEN_IF,
+                          0,
+                          0,
+                          0);
+    if (!NT_SUCCESS(Status))
+    {
+        Error = RtlNtStatusToDosError(Status);
+        SetLastError(Error);
+        return INVALID_HANDLE_VALUE;
     }
 
-    sum = (sum >> 16) + (sum & 0xffff);
-    sum  += (sum >> 16);
-    answer = ~sum;
-    return(answer);
+    return FileHandle;
 }
 
 
-
-/*
- * Exported Routines.
- */
-
 /***********************************************************************
  *		Icmp6CreateFile (IPHLPAPI.@)
  */
 HANDLE WINAPI Icmp6CreateFile(VOID)
 {
-    icmp_t* icp;
-    int sid;
-#ifdef __REACTOS__
-    WSADATA wsaData;
-
-    if (WSAStartup(MAKEWORD(2, 2), &wsaData) != ERROR_SUCCESS)
+    OBJECT_ATTRIBUTES ObjectAttributes;
+    IO_STATUS_BLOCK IoStatusBlock;
+    UNICODE_STRING DeviceName;
+    HANDLE FileHandle;
+    NTSTATUS Status;
+    DWORD Error;
+
+    /* ICMP requests are handled by the Ip device */
+    RtlInitUnicodeString(&DeviceName, L"\\Device\\Ip6");
+    InitializeObjectAttributes(&ObjectAttributes,
+                               &DeviceName,
+                               OBJ_CASE_INSENSITIVE,
+                               NULL,
+                               NULL);
+
+    /* Get a handle to the device */
+    Status = NtCreateFile(&FileHandle,
+                          GENERIC_EXECUTE,
+                          &ObjectAttributes,
+                          &IoStatusBlock,
+                          0,
+                          FILE_ATTRIBUTE_NORMAL,
+                          FILE_SHARE_READ | FILE_SHARE_WRITE,
+                          FILE_OPEN_IF,
```

Comment:
I don't think trying to create these files makes sense, so FILE_OPEN might be better.